core/src/main/java/org/apache/iceberg/avro/GenericAvroReader.java

```diff
@@ -30,13 +30,17 @@
 import org.apache.iceberg.common.DynClasses;
 import org.apache.iceberg.data.avro.DecoderResolver;
 
-class GenericAvroReader<T> implements DatumReader<T>, SupportsRowPosition {
+public class GenericAvroReader<T> implements DatumReader<T>, SupportsRowPosition {
```

Comment:
Note about this to myself for later...

This needs to use `GenericAvroReader` for a couple reasons. First, `GenericAvroReader` is the only one that supports creating records with a specific class. We could extend support so that `DataReader` can create `StructLike` instances, but that hits the second reason: using Iceberg generics will use different representations for some types, like using `LocalDate` instead of `int` for `date` type. The representation differences don't affect the `KeyMetadata` class right now, but it's still more correct to use the internal representation.